### PR TITLE
Finish midi sequencer and export

### PR DESCRIPTION
**Pull request by oakes**
This change creates a separate midi sequencer each time you play, and opens/closes them similar to how it is done with the synth. This builds off of #10

### CODE REVIEW

**Review by daveyarwood — COMMENTED** [review]
The code all looks good to me -- super exciting stuff. I'll do some testing soon and if things are feeling and sounding good, I'll merge this and get it into a release.

**Comment by daveyarwood** [comment]
Overall, this is sounding really awesome so far. I am noticing some small issues that we'll need to fix before I'm comfortable merging and releasing this. Jotting down the issues in detail here, then I'll post a tl;dr below:

- [My little hack in the Bach cello suite example](https://github.com/alda-lang/alda-core/blob/a24af57/examples/bach_cello_suite_no_1.alda#L44-L51), using 2 voices to represent a cello alternating between the same A note on multiple strings, with the open A string ringing out and overlapping each time, sounds totally weird and clipped now. I know exactly why this is happening. With the JSyn + MIDI Synthesizer implementation we have now in master, we are manually starting and stopping notes via `midi/play-note!` and `midi/stop-note!`. Because we were doing this manually instead of just scheduling NOTE ON/OFF events in a Sequencer at certain offsets, we had more flexibility about notes overlapping, so I implemented this [`protect-note!`](https://github.com/alda-lang/alda-sound-engine-clj/blob/b36f04a/src/alda/sound/midi.clj#L139-L148) thing that gets called first thing when you play a note. Essentially, it fixes issues that arise when you have multiple notes in a row on the same channel that are the same note, e.g. if you play a middle C whole note, then immediately play a middle C after it before the first note is finished -- without `protect-note!`, the second middle C could be cut off early by the NOTE OFF event of the first middle C.

  So, this is kind of a weird thing that I'd like to fix in the Sequencer-based implementation too, but I'm not quite sure how it could be done... It's sort of a general problem with MIDI sequencing that we are only currently able to work around because we're dynamically choosing whether or not to send the NOTE OFF message, on-the-fly. The Sequencer way is totally different, though... we're scheduling all of the NOTE ONs and NOTE OFFs at once and then hitting "play." I guess we could always punt and not offer this behavior, because MIDI is MIDI and there's not much we can do about that. I could rewrite that part of the Bach cello suite to use a second cello track instead of voices, and actually it would probably sound better anyway. I think I'm OK with this.

- I don't think percussion is implemented. When I try playing the [clapping music](https://github.com/alda-lang/alda-core/blob/master/examples/clapping_music.alda) and [percussion](https://github.com/alda-lang/alda-core/blob/master/examples/percussion.alda) examples, I get a NPE:

    ```
    18-04-06 02:52:44 moondog ERROR [alda.worker:108] - java.lang.NullPointerException
                                  java.lang.Thread.run              Thread.java:  748
    java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  624
     java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1149
                   java.util.concurrent.FutureTask.run          FutureTask.java:  266
                                                   ...
                   clojure.core/binding-conveyor-fn/fn                 core.clj: 1938
                       alda.worker/handle-code-play/fn               worker.clj:  131
                                  alda.worker/run-job!               worker.clj:  101
                               alda.worker/run-job!/fn               worker.clj:  101
                              alda.now/play-with-opts!                  now.clj:   79
                                      alda.sound/play!                sound.clj:  331
                          alda.sound/score-to-sequence                sound.clj:  263
    java.lang.NullPointerException:
    ```

- Function events are not implemented, but that was already a TODO item. Playing [this example score with scheduled printing](https://github.com/alda-lang/alda-core/blob/master/examples/printing.alda), for example, causes the same stacktrace as above, interestingly enough.

  I'm not sure that we even need this feature at this point. It's kind of a cool idea -- executing arbitrary user-defined Clojure functions synchronized to a certain point in the score -- but so far, I've only ever used it to print stuff, but it's basically just a novelty IMO, and it actually doesn't even work anymore since we moved to a client/server architecture, because the printing is all happening in a background process, so you can't see it. (see: https://github.com/alda-lang/alda-server-clj/issues/1)

  We will eventually need to add support for custom events so that we can use our Sequencer to do stuff like play waveform synthesis instruments, or trigger samples, etc. but that kind of thing is a long way off. I think it actually makes sense to remove the scheduled function feature from Alda and not worry about this for now.

- We need to implement panning, e.g. so [this example](https://github.com/alda-lang/alda-core/blob/master/examples/panning.alda) sounds like it's shifting from one speaker to another.

**Comment by daveyarwood** [comment]
tl;dr of what needs to happen next:

- [x] I need to update the Bach cello suite example to use a second cello instead of voices.
- [x] Remove scheduled functions feature from the code and docs.
- [x] Implement percussion / fix NPE when running the "clapping" and "percussion" examples in the alda-core repo.
- [x] Implement panning and track-volume attributes. See [play-note!](https://github.com/alda-lang/alda-sound-engine-clj/blob/b36f04a/src/alda/sound/midi.clj#L168-L179) to get an idea of what we do with those values currently. Note that there is a subtle difference between volume and track-volume. Right now we are doing the volume part (including a volume value 0-127 on each NOTE ON message), but not doing the track-volume or panning "control changes."
- [x] ~~Address "abrupt cut-off" issues (only evident in REPL-based workflow)~~ (punting on making this experience better for now. Even on master, playback is not the smoothest. I think we need to re-think how to support using Alda as a library in REPL-based workflows)
- [x] ~~Fix Sequence constructor bug: I think we should use `(new Sequence Sequence/SMTPE_24 1)` and make sure that still works.~~ (punting on this... the time conversion is confusing)

**Comment by jgkamat** [comment]
I actually accidentally changed the behavior of `load-instruments!` slightly when adding support for loading into receivers, I *think* I got it right now. Percussion seems to work fine for now (those examples sound really nice :D).

The commit should be added to this PR now \o/

**Comment by daveyarwood** [comment]
That seems to fix the NPE issue and get percussion working. Thanks @jgkamat ! :+1:

File: src/alda/sound.clj [hunk 1]
```diff
@@ -310,27 +326,15 @@
...
-  [score & [event-set]]
+  [score & args]
   (let [{:keys [one-off? async?]} *play-opts*
         _           (log/debug "Determining audio types...")
         score       (update score :audio-context #(or % (new-audio-context)))
-        _           (log/debug "Setting up audio types...")
-        _           (set-up! score)
-        _           (refresh! score)
+        sequence    (apply create-sequence! score args)
         playing?    (atom true)
```
Comment (daveyarwood) [inline src/alda/sound.clj]: I think we can get rid of this `playing?` atom now -- it's an implementation detail of the current JSyn-based system, and not used for anything anymore on this branch.

**Comment by oakes** [comment]
@daveyarwood 

I played with this quite heavily over the weekend and one thing that seems to be an issue for me is the performance of the midi sequencer when doing normal playback. I suspect this won't be an issue for a typical alda user, but when I'm doing the instaREPL workflow like in my edna video, the quick starts/stops seems to throw off the sequencer, and it seems to stutter and sometimes skip notes entirely.

As an experiment, I brought back the optional JSyn code and enabled it during playback, and it definitely solves the issue for me. I totally understand wanting to remove a dependency, and if this particular issue doesn't affect alda users, I don't mind just running off of a fork. What do you think? Of course, if there's any way we can make the sequencer perform better, that would be ideal...

**Comment by daveyarwood** [comment]
This is where I'm curious if pooling Sequencers would help us. I imagine there is a startup cost when creating a Sequencer instance. There definitely is a startup cost of at least half a second when creating a Synthesizer instance, which is what led me to make the synthesizer pool in alda.sound.midi. So I think that's definitely something we ought to try with Sequencers.

EDIT: Forgot to mention that I'd totally like to get this working for both Clojure REPL-based and Alda CLI-based workflows!

**Comment by oakes** [comment]
Not sure why I didn't try that before....it seems to do the trick! I have to play with it more but I pushed a rough attempt that basically copies your pooling code and duplicates it for sequencers. So far I haven't heard any stuttering.

**Comment by daveyarwood** [comment]
Yes, that looks like exactly what I had in mind! Well done, sir!

**Comment by oakes** [comment]
After further use (i need to go to bed...) another issue with midi sequencer playback is that it seems to stop abruptly. For example, playing just a single `:c` note on the piano with the midi sequencer sounds like it's being cut off at the end. This doesn't seem to be the case with the exported version, though.

**Comment by daveyarwood** [comment]
Hmm... without having time to dig into it at the moment, that sounds like we're either closing the Sequencer prematurely, or perhaps the "end of track" metamessage is being placed too early in the sequence?

[Here's where we're sending the "end of track" metamessage](https://github.com/oakes/alda-sound-engine-clj/blob/0612fe5526735a6a7835032752c44ce4ed2706d7/src/alda/sound.clj#L281-L284), but I don't really understand where that gets placed, off-hand... maybe the -1 is significant?

**Comment by jgkamat** [comment]
When I tried adding the `END_OF_TRACK` message, I don't think it made any difference the point when I added the event at, it seemed to position itself at the end. I'm not 100% sure about this though.

Because of this, I followed this from the docs:

> If time-stamping is not supported by this receiver, the time-stamp value should be -1

Even though time-stamping was supported by the receiver. I'll take another look at playing with the end of track message when I get the time, the docs aren't clear enough how it works :(

File: src/alda/sound.clj [hunk 2]
```diff
@@ -251,35 +241,83 @@
...
-                                     (/ offset 1000.0)
-                                     (/ duration 1000.0)) stop!))))
-    (schedule-event! engine (+ begin
-                               (/ (score-length events) 1000.0)
-                               1) end!)))
+        {:keys [midi-synth midi-channels]} @audio-context
+        channels  (.getChannels ^Synthesizer midi-synth)
+
+        ;; TODO make resolution configurable
+        seq (new Sequence Sequence/PPQ Sequence/SMPTE_24)
```
Comment (daveyarwood) [inline src/alda/sound.clj]: I'm reading up on Sequences now, and I think we may have a bug here.

The [Sequence constructor](https://docs.oracle.com/javase/7/docs/api/javax/sound/midi/Sequence.html#Sequence(float,%20int)) being invoked here takes 2 arguments:

1. (float) the timing division type (PPQ or one of the SMPTE types)
2. (int) the timing resolution

I think we need to choose either PPQ or one of the SMPTE types as the first argument, and then specify the resolution as the second argument. If we choose PPQ, then the second argument is how many ticks per quarter note, and if we choose SMPTE, the second argument is how many ticks per SMTPE frame, which effectively makes it a coefficient -- e.g. if we're using SMTPE_24, then that's 24 frames per second, and then if resolution is 2, then there are 48 ticks per second.

It looks like `Sequence/SMPTE_24` has a value of 24.0:

```clojure
boot.user=> (import '[javax.sound.midi Sequence])
javax.sound.midi.Sequence
boot.user=> Sequence/SMPTE_24
24.0
```

So, I think that means we're getting a resolution of 24 ticks per quarter note, which actually is not terrible, depending on the tempo, but I think it would be better for us to specify a sensible resolution. Given that Alda lets you use crazy non-standard rhythms specified in seconds/milliseconds, it would probably be better for us to use one of the SMTPE time codes instead of PPQ. I think `(new Sequence Sequence/SMTPE_24 1)` is probably sensible for now, and we can tweak it in the future if we ever need to for whatever reason.

Comment (daveyarwood) [reply src/alda/sound.clj]: Showing my math: SMTPE_24 at a resolution of 1 would be 24 ticks per second (24 frames per second, 1 tick per frame). At 60 bpm (1 quarter note beat per second), timing could be accurate up to 24 notes per beat, which I think is way more granular than any composer would ever need. You can fit sixteen 64th notes into the span of a quarter note, and you rarely ever see 64th notes in written scores, and if you do, it's going to be at a very slow tempo, almost certainly slower than 60 bpm because 32nd notes are already very fast at 60 bpm. And that's just 64th notes. 24 notes per beat is the equivalent of 96th notes (i.e in Alda, you could write `c96 d e f`)!

File: src/alda/sound.clj [hunk 3]
```diff
@@ -251,35 +241,83 @@
...
-                                   (/ offset 1000.0)) start!)
-        (when-not (:function event)
-          (schedule-event! engine (+ begin
-                                     (/ offset 1000.0)
-                                     (/ duration 1000.0)) stop!))))
-    (schedule-event! engine (+ begin
-                               (/ (score-length events) 1000.0)
-                               1) end!)))
+        {:keys [midi-synth midi-channels]} @audio-context
+        channels  (.getChannels ^Synthesizer midi-synth)
```
Comment (daveyarwood) [inline src/alda/sound.clj]: This line can be removed -- `channels` is not used below.

**Comment by oakes** [comment]
@daveyarwood I haven't narrowed it down yet, but using `(new Sequence Sequence/SMPTE_24 1)` locally and it seems to break exporting. The resulting midi file works (though plays at a higher tempo) but oddly enough when I convert it to wav or mp3 it takes a long time, and then produces a file that plays just a single note. This could be a problem with my wav/mp3 conversion code I suppose...

**Comment by daveyarwood** [comment]
@oakes Alright, I went a little further down the rabbit hole and I think the issue here is that MidiEvent timestamps are expressed in _ticks_ instead of microseconds. The Sequencer's clock is in microseconds, but it can handle any number of Sequences that can each have a different division type, so _ticks_ are sort of a division-type-agnostic way to express a timestamp on a MIDI event.

Alda deals with time uniquely in that events have no notion of tempo; they just have an absolute offset in milliseconds.

So, I think we have to translate our millisecond offsets into ticks, whereas at the moment (on this branch), we're multiplying the milliseconds by 1000 to get microseconds. If there are 24 ticks per second, then an offset of 1 second (1000 ms) would be 24 ticks. Therefore, I think the formula is `(offset / 1000) * 24`? I'm not 100% that's correct, but that's my current understanding, based on what I've read. Maybe worth a try to do `(offset / 1000) * 24` instead of `offset * 1000` and seeing if that fixes it.

**Comment by oakes** [comment]
I am still having some problems. I created a project to reproduce it:

https://github.com/oakes/alda-export-example

If you install `midi-export-and-remove-jsyn-wip` locally and run `boot build` in that project, you should get an mp3 in your target folder that sounds correct. But if you make the Sequencer and offset changes and then reinstall, it will sound like one note being held for a long time. Let me know if you have any trouble with the project...I just made it and i'm drinking bourbon, so mistakes are possible lol.

**Comment by daveyarwood** [comment]
I'll give that a try soon, though I'm thinking that if we start getting too far out in the weeds with this, it might make sense for us to just go back to `(new Sequence Sequence/PPQ Sequence/SMPTE_24)` and leave it alone for now, since that's clearly working! (even if it may be technically incorrect)

**Comment by oakes** [comment]
Cool, let me know if there's anything else.

**Comment by daveyarwood** [comment]
Looking at my punch-list [above](https://github.com/alda-lang/alda-sound-engine-clj/pull/11#issuecomment-379139795), I think all we're missing that's relevant to this repo is:

- Implement panning and track-volume attributes. See [play-note!](https://github.com/alda-lang/alda-sound-engine-clj/blob/b36f04a/src/alda/sound/midi.clj#L168-L179) to get an idea of what we do with those values currently. Note that there is a subtle difference between volume and track-volume. Right now we are doing the volume part (including a volume value 0-127 on each NOTE ON message), but not doing the track-volume or panning "control changes."
- Address "abrupt cut-off" issues (only evident in REPL-based workflow?)

@oakes Are you still seeing the abrupt cut-off issues?

**Comment by oakes** [comment]
Ah yes, it appears the abrupt cut-off is still happening. In fact, my example project above makes it easy to reproduce. Install this branch locally. Then in the example project, change the `music` var in `alda-export-example.core` to just be `(def music [:piano :c])`. I find it is easier to notice when you're just playing a single note. Then either do `boot run` to play it once or fire up boot's REPL and do `(require '[alda-export-example.core :refer [-main]])` so you can just run `(-main)` to play the note (though for some reason the first time you run it you can't hear anything...probably something dumb that i did).

BTW that project is using a vendored copy of edna which might make it easier to mess with it. I'd be surprised if this issue was something specific to it, though it's possible. You can always use my `edna->alda` function to get the underlying alda data that it's playing.

**Comment by daveyarwood** [comment]
I played around with this for a while tonight. Here are my findings (this is a bit long-winded, bear with me!)

@oakes: I tried your example project and I was able to reproduce the issue with the score ending too early -- this was especially noticeable with the one-note score `piano: c`, but I could also tell that the last chord in the excerpt of Aeris' theme was being cut off instead of fully tapering off.

I was able to come up with a more minimal repro of the issue, ruling out anything you're doing with edna as the cause -- I've added that to the master branch of this repo in the form of an `alda.sound.dev` namespace with some commented out scratch code -- [see here](https://github.com/alda-lang/alda-sound-engine-clj/commit/73a8610f03b8aeb8acfd7d7bc2774c2578a67a21). This sounds OK-ish* on the master branch, but if you switch to this branch, add `"test"` to the source-paths, and try the same thing, the Sequencer-based implementation cuts off the score early.

I did try this, to see if maybe it would fix the issue, under the assumption that the issue is that the "end of track" message was being scheduled too early:

```clojure
    ;; auto-add a end of track metamessage...
    (.send receiver
           (doto (new MetaMessage)
             (.setMessage midi/MIDI-END-OF-TRACK nil 0))
           ;; ...just past the end of the score.
           (+ (score-length events) 1000))
```

That didn't seem to have any bearing on the issue, though. In fact, I commented out that entire form and it didn't seem to make any difference at all.

*In general, I'm finding that Alda isn't playing very smoothly in REPL-based workflows, either on this branch or on master. Especially when playing one thing right after another, the audio for me is often very choppy, cutting in and out, ending early, or starting late and jumping ahead. I think this might have something to do with how we're using multiple Synthesizer instances in a throw-away fashion. These Synthesizers are fairly heavyweight, and the fact that we're doing all of the score processing as well as the sound generation in the same process is not helping at all. I read that this is super important when you're doing audio programming in general -- sound generation should be done in a separate thread or ideally a separate process, so that it doesn't have to compete for compute cycles with other threads/processes. Apparently that's the key to having uninterrupted, well-timed playback.

Running Alda from the command line results in playback that is a lot smoother, and I think that's because an Alda worker process is only using one Synthesizer at a time. It opens the Synthesizer when it starts up, and while it's playing a score, it will refuse to play another score until it's done with the current score or you tell it to stop. Workers also kill themselves off after a certain period of time, and the server spins up new ones, ensuring that fresh workers are available all the time, each one a separate process using only a single Synthesizer. I've had a lot of success doing things that way (oriented around simple processes talking to each other, instead of one big, complicated process using lots of threads), and I have plans to push further in that direction, making the processes that do the playback even lighter-weight and having the server take on more of the data crunching work.

I have some thinking to do about to how to support REPL workflows / using Alda as a library from Clojure projects like edna. I definitely think that's really important, but I think I might want to do it a bit differently. I'll mull that one over.

To be clear, I do want to move forward with this branch and start using Sequencers for playback. I built `alda` locally using this branch and https://github.com/alda-lang/alda-server-clj/pull/11 and played around with it on the command-line, and aside from the remaining checklist items (panning and track-volume need to be implemented, and I think that's it!), I think it's working well, at least well enough to be fit for command-line Alda usage. And it will get us a lot closer to being able to export MIDI, which would be amazing.

**Comment by daveyarwood** [comment]
A little update on this issue:

I've been hacking on this branch a little bit at a time, and I think I almost have it to a point where it's merge-worthy. I found and fixed a couple of bugs. Most significantly, I realized that when you call `Sequencer.startRecording`, it doesn't only start recording, it starts playback also! Which means that when we start piping events into the sequencer's receiver, they'll start playing back immediately, which can result in a jittery start as the sequencer and synthesizer struggle to catch up with the flurry of scheduled notes and begin to play them in time.

I found a someone's Java example of another way to do it that I think maps more closely to what we're trying to do in Alda. You can call `Track.add` on the sequence track, once for each event, and add them to the track instead of starting to record and then sending messages to the receiver. `Track.add` takes a MidiEvent, which is just a wrapper for a MidiMessage (the kind you would send to a receiver) and an offset in ticks.

(Through experimentation, I was able to make sense of the confusing milliseconds to ticks conversion. Basically, I think I was right that a `Sequence(SMTPE_24, 1)` will work, and you have to convert milliseconds -> seconds -> ticks, which is as simple as dividing by 1000 and multiplying by 24.)

After all the events have been added to the track, all we have to do is set the position to 0 ticks and call `Sequencer.play` and it plays the events in the sequence from the beginning.

I still have a couple of tweaks to make, but I'll get this branch updated soon and see where we're at with being able to merge it.

**Comment by daveyarwood** [comment]
I wasn't sure how to link this commit here, but I'm working on a separate branch. I made a WIP commit just now: ee7b635

Still a handful of issues to fix. Will keep hacking!

**Comment by daveyarwood** [comment]
I did some more work on this tonight: https://github.com/alda-lang/alda-sound-engine-clj/commit/aaa9c0b34cc9874f404b767aadd324edc54960ef, https://github.com/alda-lang/alda-sound-engine-clj/commit/5fbca8406a371c962b7a6ff6cf552098030ff268

I'm particularly excited about my discovery in that second commit. The timing gets _way_ more solid when we use SMTPE_24 with 2 ticks per frame instead of 1 tick per frame. That's 48 ticks per second instead of 24, which allows us to be a lot more accurate with the timing of the notes.

The last issue that I need to chase down is that I'm noticing some weirdness around instrument patches not changing right away when playing one score (with different instruments) after another, as well as some issues with what sounds like multiple synthesizers playing the same score at once. I think I probably have some bugs in the code that manages the lifetimes of the synthesizers, sequencers, transmitters and receivers. I'd like to clean up that aspect of the code a little, anyway.

Getting closer!

**Comment by TBuc** [comment]
Great work Dave!

Some food for thought: does this maybe mean that Fourier stating that the
sampling frequency shall be at least double the source frequency in order
for the conversion to be lossless, might apply to this topic too?

Il Dom 23 Set 2018, 05:08 Dave Yarwood <notifications@github.com> ha
scritto:

**Comment by daveyarwood** [comment]
@TBuc I think that is relevant here. I believe the issue here does have to something to do with lossiness.

Alda allows you to express rhythms with a totally arbitrary level of precision. You can choose to write a piece of music that has a very fast tempo and uses very short note durations (64th notes, 128th notes, etc.), and the placement of each note is then converted into a millisecond timestamp (and the length of the note is also converted into milliseconds). Then, when it comes time to play the score, we have to convert all of those millisecond values into _ticks_, which are less precise. So, we do end up losing some information about exactly how a note should be placed if it happens not to line up with a tick exactly and we have to round up/down. The trade-off, of course, is performance -- the higher tick resolution we choose, the more demand that puts on your system.

@jgkamat wrote a TODO comment in the code that we might consider making the resolution configurable -- I think that would be a good idea at some point. For now, I've chosen a modest value that sounds good on my laptop and I'm going to hope for the best... :sweat_smile:

**Comment by daveyarwood** [comment]
OK, I think we're just about ready to test and hopefully release this! I've made a PR from my new branch: #14 

Let's continue the conversation there.
